pkg/buildpod/generator.go

```diff
@@ -84,13 +91,64 @@ func (g *Generator) buildAllowed(ctx context.Context, build BuildPodable) error
 		}
 	}
 
-	for _, binding := range build.Bindings() {
-		if binding.SecretRef != nil && forbiddenSecrets[binding.SecretRef.Name] {
-			return fmt.Errorf("binding %q uses forbidden secret %q", binding.Name, binding.SecretRef.Name)
+	bindings := make([]buildapi.ServiceBinding, 0)
+
+	if services := build.Services(); len(services) != 0 {
+		for _, s := range build.Services() {
+			var b corev1alpha1.ServiceBinding
+			if s.Kind == "Secret" {
+				b = corev1alpha1.ServiceBinding{
+					Name:      s.Name,
+					SecretRef: &corev1.LocalObjectReference{Name: s.Name},
+				}
+			} else {
+				gvr, _ := meta.UnsafeGuessKindToResource(s.GroupVersionKind())
+				unstructured, err := g.DynamicClient.Resource(gvr).Namespace(build.GetNamespace()).Get(ctx, s.Name, metav1.GetOptions{})
+				if err != nil {
+					return nil, err
+				}
+				ps := duckprovisionedserviceable.ProvisionedServicable{}
+				if err := duck.FromUnstructured(unstructured, &ps); err != nil {
+					return nil, err
+				}
+				b = corev1alpha1.ServiceBinding{
+					Name:      s.Name,
+					SecretRef: ps.Status.Binding,
+				}
+			}
+			if forbiddenSecrets[b.SecretRef.Name] {
+				return nil, fmt.Errorf("build rejected: service %q uses forbidden secret %q", b.Name, b.SecretRef.Name)
+			}
+
+			secret, err := g.K8sClient.CoreV1().Secrets(build.GetNamespace()).Get(ctx, b.SecretRef.Name, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+
+			if t, ok := secret.StringData["type"]; !ok || secret.Type == "" || fmt.Sprintf("service.binding/%s", t) != string(secret.Type) {
```

Comment:
Do we actually care?

Reply:
are you thinking we just let it through and let the build potentially fail?